File: BackEnd/PHP/laravelshop/app/Modules/Friend/Services/FriendService.php
<?php

namespace App\Modules\Friend\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\FriendActivity;
use App\Models\FriendBlock;
use App\Models\Friendship;
use App\Models\User;
use Symfony\Component\HttpFoundation\Response;

class FriendService
{
    public function request(User $actor, int $userId): array
    {
        if ($actor->id === $userId) {
            throw new BusinessException('본인에게 친구 요청을 보낼 수 없습니다.', 'FRIEND_REQUEST_INVALID', Response::HTTP_BAD_REQUEST);
        }

        $target = User::query()->find($userId);
        if (! $target) {
            throw new BusinessException('사용자를 찾을 수 없습니다.', 'USER_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        if ($this->isBlocked($actor->id, $userId) || $this->isBlocked($userId, $actor->id)) {
            throw new BusinessException('차단 관계에서는 친구 요청을 보낼 수 없습니다.', 'FRIEND_BLOCKED', Response::HTTP_BAD_REQUEST);
        }

        Friendship::query()->firstOrCreate([
            'requester_id' => $actor->id,
            'addressee_id' => $userId,
        ], [
            'status' => 'PENDING',
        ]);

        return ['message' => '친구 요청을 보냈습니다.'];
    }

    public function accept(User $actor, int $friendshipId): array
    {
        $friendship = Friendship::query()->find($friendshipId);
        if (! $friendship || $friendship->addressee_id !== $actor->id) {
            throw new BusinessException('친구 요청을 찾을 수 없습니다.', 'FRIEND_REQUEST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $friendship->forceFill(['status' => 'ACCEPTED'])->save();
        FriendActivity::query()->create([
            'user_id' => $actor->id,
            'type' => 'FRIEND_ACCEPTED',
            'message' => '친구 요청을 수락했습니다.',
        ]);

        return ['message' => '친구 요청을 수락했습니다.'];
    }

    public function reject(User $actor, int $friendshipId): array
    {
        $friendship = Friendship::query()->find($friendshipId);
        if (! $friendship || $friendship->addressee_id !== $actor->id) {
            throw new BusinessException('친구 요청을 찾을 수 없습니다.', 'FRIEND_REQUEST_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $friendship->forceFill(['status' => 'REJECTED'])->save();

        return ['message' => '친구 요청을 거절했습니다.'];
    }

    public function list(User $actor, int $page, int $limit): array
    {
        $result = Friendship::query()
            ->where('status', 'ACCEPTED')
            ->where(function ($q) use ($actor): void {
                $q->where('requester_id', $actor->id)->orWhere('addressee_id', $actor->id);
            })
            ->orderByDesc('id')
            ->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (Friendship $friendship): array => $this->serializeFriendship($actor, $friendship))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function received(User $actor, int $page, int $limit): array
    {
        $result = Friendship::query()->where('addressee_id', $actor->id)->where('status', 'PENDING')->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (Friendship $friendship): array => $this->serializeRequest($friendship))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function sent(User $actor, int $page, int $limit): array
    {
        $result = Friendship::query()->where('requester_id', $actor->id)->where('status', 'PENDING')->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (Friendship $friendship): array => $this->serializeRequest($friendship))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function feed(User $actor, int $page, int $limit): array
    {
        $friendIds = Friendship::query()
            ->where('status', 'ACCEPTED')
            ->where(function ($q) use ($actor): void {
                $q->where('requester_id', $actor->id)->orWhere('addressee_id', $actor->id);
            })
            ->get()
            ->flatMap(fn (Friendship $f) => [$f->requester_id, $f->addressee_id])
            ->filter(fn ($id) => $id !== $actor->id)
            ->unique()
            ->values();

        $result = FriendActivity::query()->whereIn('user_id', $friendIds)->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (FriendActivity $activity): array => [
                'id' => $activity->id,
                'userId' => $activity->user_id,
                'type' => $activity->type,
                'message' => $activity->message,
                'createdAt' => optional($activity->created_at)?->toISOString(),
            ])->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function block(User $actor, int $userId): array
    {
        FriendBlock::query()->firstOrCreate([
            'user_id' => $actor->id,
            'blocked_user_id' => $userId,
        ]);

        Friendship::query()
            ->where(function ($q) use ($actor, $userId): void {
                $q->where('requester_id', $actor->id)->where('addressee_id', $userId);
            })
            ->orWhere(function ($q) use ($actor, $userId): void {
                $q->where('requester_id', $userId)->where('addressee_id', $actor->id);
            })
            ->delete();

        return ['message' => '사용자를 차단했습니다.'];
    }

    public function unblock(User $actor, int $userId): array
    {
        FriendBlock::query()->where('user_id', $actor->id)->where('blocked_user_id', $userId)->delete();

        return ['message' => '사용자 차단을 해제했습니다.'];
    }

    public function remove(User $actor, int $userId): array
    {
        Friendship::query()
            ->where('status', 'ACCEPTED')
            ->where(function ($q) use ($actor, $userId): void {
                $q->where('requester_id', $actor->id)->where('addressee_id', $userId);
            })
            ->orWhere(function ($q) use ($actor, $userId): void {
                $q->where('requester_id', $userId)->where('addressee_id', $actor->id);
            })
            ->delete();

        return ['message' => '친구를 삭제했습니다.'];
    }

    private function isBlocked(int $userId, int $targetId): bool
    {
        return FriendBlock::query()->where('user_id', $userId)->where('blocked_user_id', $targetId)->exists();
    }

    private function serializeFriendship(User $actor, Friendship $friendship): array
    {
        $friendId = $friendship->requester_id === $actor->id ? $friendship->addressee_id : $friendship->requester_id;
        $friend = User::query()->find($friendId);

        return [
            'id' => $friendship->id,
            'friend' => $friend ? [
                'id' => $friend->id,
                'name' => $friend->name,
                'nickname' => $friend->nickname,
            ] : null,
            'status' => $friendship->status,
            'createdAt' => optional($friendship->created_at)?->toISOString(),
        ];
    }

    private function serializeRequest(Friendship $friendship): array
    {
        return [
            'id' => $friendship->id,
            'requesterId' => $friendship->requester_id,
            'addresseeId' => $friendship->addressee_id,
            'status' => $friendship->status,
            'createdAt' => optional($friendship->created_at)?->toISOString(),
        ];
    }

    private function pagination($result): array
    {
        return [
            'page' => $result->currentPage(),
            'limit' => $result->perPage(),
            'total' => $result->total(),
            'totalPages' => $result->lastPage(),
        ];
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/friend/FriendService.java
package com.pbshop.java.spring.maven.jpa.postgresql.friend;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.user.User;
import com.pbshop.java.spring.maven.jpa.postgresql.user.UserRepository;

@Service
@Transactional
public class FriendService {
    private final FriendshipRepository friendshipRepository;
    private final FriendBlockRepository friendBlockRepository;
    private final FriendActivityRepository friendActivityRepository;
    private final UserRepository userRepository;

    public FriendService(FriendshipRepository friendshipRepository, FriendBlockRepository friendBlockRepository,
            FriendActivityRepository friendActivityRepository, UserRepository userRepository) {
        this.friendshipRepository = friendshipRepository;
        this.friendBlockRepository = friendBlockRepository;
        this.friendActivityRepository = friendActivityRepository;
        this.userRepository = userRepository;
    }

    public Map<String, Object> requestFriend(AuthenticatedUserPrincipal principal, Long userId) {
        User actor = requireCurrentUser(principal);
        if (actor.getId().equals(userId)) {
            throw new BusinessException(ErrorCode.BAD_REQUEST, "본인에게 친구 요청을 보낼 수 없습니다.");
        }
        User target = requireUser(userId);
        if (friendBlockRepository.existsByUserIdAndBlockedUserId(actor.getId(), userId)
                || friendBlockRepository.existsByUserIdAndBlockedUserId(userId, actor.getId())) {
            throw new BusinessException(ErrorCode.BAD_REQUEST, "차단 관계에서는 친구 요청을 보낼 수 없습니다.");
        }
        friendshipRepository.findByRequesterIdAndAddresseeId(actor.getId(), target.getId()).orElseGet(() -> {
            Friendship friendship = new Friendship();
            friendship.setRequester(actor);
            friendship.setAddressee(target);
            friendship.setStatus("PENDING");
            return friendshipRepository.save(friendship);
        });
        return Map.of("message", "친구 요청을 보냈습니다.");
    }

    public Map<String, Object> accept(AuthenticatedUserPrincipal principal, Long friendshipId) {
        User actor = requireCurrentUser(principal);
        Friendship friendship = findFriendship(friendshipId);
        if (!friendship.getAddressee().getId().equals(actor.getId())) {
            throw new BusinessException(ErrorCode.NOT_FOUND, "친구 요청을 찾을 수 없습니다.");
        }
        friendship.setStatus("ACCEPTED");
        friendActivityRepository.save(activity(actor, "FRIEND_ACCEPTED", "친구 요청을 수락했습니다."));
        return Map.of("message", "친구 요청을 수락했습니다.");
    }

    public Map<String, Object> reject(AuthenticatedUserPrincipal principal, Long friendshipId) {
        User actor = requireCurrentUser(principal);
        Friendship friendship = findFriendship(friendshipId);
        if (!friendship.getAddressee().getId().equals(actor.getId())) {
            throw new BusinessException(ErrorCode.NOT_FOUND, "친구 요청을 찾을 수 없습니다.");
        }
        friendship.setStatus("REJECTED");
        return Map.of("message", "친구 요청을 거절했습니다.");
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getFriends(AuthenticatedUserPrincipal principal) {
        User actor = requireCurrentUser(principal);
        List<Map<String, Object>> items = friendshipRepository.findByStatusOrderByIdDesc("ACCEPTED").stream()
                .filter(friendship -> friendship.getRequester().getId().equals(actor.getId())
                        || friendship.getAddressee().getId().equals(actor.getId()))
                .map(friendship -> toFriendResponse(actor, friendship))
                .toList();
        return Map.of("items", items, "pagination", pagination(items.size()));
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getReceived(AuthenticatedUserPrincipal principal) {
        User actor = requireCurrentUser(principal);
        List<Map<String, Object>> items = friendshipRepository.findByAddresseeIdAndStatusOrderByIdDesc(actor.getId(), "PENDING")
                .stream().map(this::toRequestResponse).toList();
        return Map.of("items", items, "pagination", pagination(items.size()));
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getSent(AuthenticatedUserPrincipal principal) {
        User actor = requireCurrentUser(principal);
        List<Map<String, Object>> items = friendshipRepository.findByRequesterIdAndStatusOrderByIdDesc(actor.getId(), "PENDING")
                .stream().map(this::toRequestResponse).toList();
        return Map.of("items", items, "pagination", pagination(items.size()));
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getFeed(AuthenticatedUserPrincipal principal) {
        User actor = requireCurrentUser(principal);
        List<Long> friendIds = friendshipRepository.findByStatusOrderByIdDesc("ACCEPTED").stream()
                .filter(friendship -> friendship.getRequester().getId().equals(actor.getId())
                        || friendship.getAddressee().getId().equals(actor.getId()))
                .flatMap(friendship -> List.of(friendship.getRequester().getId(), friendship.getAddressee().getId()).stream())
                .filter(id -> !id.equals(actor.getId()))
                .distinct()
                .toList();
        List<Map<String, Object>> items = friendIds.isEmpty() ? List.of()
                : friendActivityRepository.findByUserIdInOrderByIdDesc(friendIds).stream().map(this::toActivityResponse).toList();
        return Map.of("items", items, "pagination", pagination(items.size()));
    }

    public Map<String, Object> block(AuthenticatedUserPrincipal principal, Long userId) {
        User actor = requireCurrentUser(principal);
        User blocked = requireUser(userId);
        friendBlockRepository.findByUserIdAndBlockedUserId(actor.getId(), blocked.getId()).orElseGet(() -> {
            FriendBlock block = new FriendBlock();
            block.setUser(actor);
            block.setBlockedUser(blocked);
            return friendBlockRepository.save(block);
        });
        friendshipRepository.findByRequesterIdAndAddresseeId(actor.getId(), blocked.getId()).ifPresent(friendshipRepository::delete);
        friendshipRepository.findByRequesterIdAndAddresseeId(blocked.getId(), actor.getId()).ifPresent(friendshipRepository::delete);
        return Map.of("message", "사용자를 차단했습니다.");
    }

    public Map<String, Object> unblock(AuthenticatedUserPrincipal principal, Long userId) {
        User actor = requireCurrentUser(principal);
        FriendBlock block = friendBlockRepository.findByUserIdAndBlockedUserId(actor.getId(), userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "차단 대상을 찾을 수 없습니다."));
        friendBlockRepository.delete(block);
        return Map.of("message", "사용자 차단을 해제했습니다.");
    }

    public Map<String, Object> remove(AuthenticatedUserPrincipal principal, Long userId) {
        User actor = requireCurrentUser(principal);
        friendshipRepository.findByRequesterIdAndAddresseeId(actor.getId(), userId).ifPresent(friendshipRepository::delete);
        friendshipRepository.findByRequesterIdAndAddresseeId(userId, actor.getId()).ifPresent(friendshipRepository::delete);
        return Map.of("message", "친구를 삭제했습니다.");
    }

    private Friendship findFriendship(Long id) {
        return friendshipRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "친구 요청을 찾을 수 없습니다."));
    }

    private FriendActivity activity(User user, String type, String message) {
        FriendActivity activity = new FriendActivity();
        activity.setUser(user);
        activity.setType(type);
        activity.setMessage(message);
        return activity;
    }

    private Map<String, Object> toFriendResponse(User actor, Friendship friendship) {
        User friend = friendship.getRequester().getId().equals(actor.getId()) ? friendship.getAddressee() : friendship.getRequester();
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", friendship.getId());
        response.put("friend", Map.of("id", friend.getId(), "name", friend.getName(), "nickname", friend.getName()));
        response.put("status", friendship.getStatus());
        response.put("createdAt", friendship.getCreatedAt() == null ? null : friendship.getCreatedAt().toString());
        return response;
    }

    private Map<String, Object> toRequestResponse(Friendship friendship) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", friendship.getId());
        response.put("requesterId", friendship.getRequester().getId());
        response.put("addresseeId", friendship.getAddressee().getId());
        response.put("status", friendship.getStatus());
        response.put("createdAt", friendship.getCreatedAt() == null ? null : friendship.getCreatedAt().toString());
        return response;
    }

    private Map<String, Object> toActivityResponse(FriendActivity activity) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", activity.getId());
        response.put("userId", activity.getUser().getId());
        response.put("type", activity.getType());
        response.put("message", activity.getMessage());
        response.put("createdAt", activity.getCreatedAt() == null ? null : activity.getCreatedAt().toString());
        return response;
    }

    private Map<String, Object> pagination(int total) {
        return Map.of("page", 1, "limit", total == 0 ? 20 : total, "total", total, "totalPages", total == 0 ? 0 : 1);
    }

    private User requireCurrentUser(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }
        return requireUser(principal.userId());
    }

    private User requireUser(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "사용자를 찾을 수 없습니다."));
    }
}
